scapy/contrib/automotive/someip_sd.py

```diff
@@ -0,0 +1,342 @@
+    TYPE_EVTGRP_SUBSCRIBE_ACK = 0x07
+    TYPE_EVTGRP = (TYPE_EVTGRP_SUBSCRIBE, TYPE_EVTGRP_SUBSCRIBE_ACK)
+    OVERALL_LEN = 16
+
+    fields_desc = [
+        ByteField("type", 0),
+        ByteField("index_1", 0),
+        ByteField("index_2", 0),
+        BitField("n_opt_1", 0, 4),
+        BitField("n_opt_2", 0, 4),
+        ShortField("srv_id", 0),
+        ShortField("inst_id", 0),
+        ByteField("major_ver", 0),
+        X3BytesField("ttl", 0)]
+
+    def guess_payload_class(self, payload):
+        pl_type = orb(payload[_SDEntry.TYPE_PAYLOAD_I])
+
+        if (pl_type in _SDEntry.TYPE_SRV):
+            return (SDEntry_Service)
+        elif (pl_type in _SDEntry.TYPE_EVTGRP):
+            return (SDEntry_EventGroup)
+
+
+class SDEntry_Service(_SDEntry):
+    _defaults = {"type": _SDEntry.TYPE_SRV_FINDSERVICE}
+
+    name = "Service Entry"
+    fields_desc = [
+        _SDEntry,
+        IntField("minor_ver", 0)]
+
+
+class SDEntry_EventGroup(_SDEntry):
+    _defaults = {"type": _SDEntry.TYPE_EVTGRP_SUBSCRIBE}
+
+    name = "Eventgroup Entry"
+    fields_desc = [
+        _SDEntry,
+        BitField("res", 0, 12),
+        BitField("cnt", 0, 4),
+        ShortField("eventgroup_id", 0)]
+
+
+# SD Option
+#  - Configuration
+#  - LoadBalancing
+# - IPv4 EndPoint
+# - IPv6 EndPoint
+
+# - IPv4 MultiCast
+# - IPv6 MultiCast
+# - IPv4 EndPoint
+# - IPv6 EndPoint
+class _SDOption(_SDPacketBase):
+    CFG_TYPE = 0x01
+    CFG_OVERALL_LEN = 4
+    LOADBALANCE_TYPE = 0x02
+    LOADBALANCE_LEN = 0x05
+    LOADBALANCE_OVERALL_LEN = 8
+    IP4_ENDPOINT_TYPE = 0x04
+    IP4_ENDPOINT_LEN = 0x0009
+    IP4_MCAST_TYPE = 0x14
+    IP4_MCAST_LEN = 0x0009
+    IP4_SDENDPOINT_TYPE = 0x24
+    IP4_SDENDPOINT_LEN = 0x0009
+    IP4_OVERALL_LEN = 12
+    IP6_ENDPOINT_TYPE = 0x06
+    IP6_ENDPOINT_LEN = 0x0015
+    IP6_MCAST_TYPE = 0x16
+    IP6_MCAST_LEN = 0x0015
+    IP6_SDENDPOINT_TYPE = 0x26
+    IP6_SDENDPOINT_LEN = 0x0015
+    IP6_OVERALL_LEN = 24
+
+    def guess_payload_class(self, payload):
+        pl_type = orb(payload[2])
+
+        if (pl_type == _SDOption.CFG_TYPE):
+            return (SDOption_Config)
+        elif (pl_type == self.LOADBALANCE_TYPE):
+            return (SDOption_LoadBalance)
+        elif (pl_type == self.IP4_ENDPOINT_TYPE):
+            return (SDOption_IP4_EndPoint)
+        elif (pl_type == self.IP4_MCAST_TYPE):
+            return (SDOption_IP4_Multicast)
+        elif (pl_type == self.IP4_SDENDPOINT_TYPE):
+            return (SDOption_IP4_SD_EndPoint)
+        elif (pl_type == self.IP6_ENDPOINT_TYPE):
+            return (SDOption_IP6_EndPoint)
+        elif (pl_type == self.IP6_MCAST_TYPE):
+            return (SDOption_IP6_Multicast)
+        elif (pl_type == self.IP6_SDENDPOINT_TYPE):
+            return (SDOption_IP6_SD_EndPoint)
+
+
+class _SDOption_Header(_SDOption):
+    fields_desc = [
+        ShortField("len", None),
+        ByteField("type", 0),
+        ByteField("res_hdr", 0)]
```

Comment:
Can you put the closing bracket on the next line? (same for the other classes below)

Reply:
Yes, done.